feat(Forms): merge FieldGroup into FieldBlock (#2645)
dnbexperience/eufemia: packages/dnb-design-system-portal/src/docs/uilib/extensions/forms/create-component.mdx

---

import ListBasisAPIs from './create-component/ListBasisAPIs'
import {
  CreateBasicFieldComponent,
  CreateComposedFieldComponent,
} from './Examples'

# Create your own component

Eufemia Forms contains helper fields and tools so you can declaratively create interactive form components that flawlessly integrates between existing data and your custom form components.

By using the building blocks for field components, you save development time, and at the same time ensure that local, custom components work similarly, and fit into the setup with [the standardized field components](/uilib/extensions/forms/base-fields/).

```jsx
import {
  DataContext,
  Field,
  FieldBlock,
  Iterate,
  Value,
  ValueBlock,
  Visibility,
  useDataValue,
} from '@dnb/eufemia/extensions/forms'
```

> Here is an example of a custom component. Notice how the props received by your field component are passed through the `useDataValue` hook. This hook does not change the API of the props, so the props returned by the hook share the same typescript type with which it was called. However, it adds a few additional properties to simplify the standardization of field behavior. This in the form of the handler functions `handleFocus`, `handleChange` and `handleBlur`. Even if field components externally have these callback functions named with "on" (eg "onChange"), these will remain untouched, while the "handle" variants add handling that saves you a lot of extra work.

Author: I was trying to explain it by saying that the "on*" are untouched, and the component code need to call "handle*" to trigger tue `useDataValue` features (like connecting to DataContext and handling error state). More detailed explanations can be found in the code comments in `useDataValue.ts`

When you call these three functions from your own implementation of the user experience for the field component, as shown below, a lot will happen in the background. All available validation functions will be called at the right time, changes in value will be synchronized with any surrounding `DataContext`, co-operation between several fields that should display error messages collectively instead of individually, and not least it ensures that error messages are not displayed on unnecessary times such as while the user is making changes to the field.

<CreateBasicFieldComponent />

In the example above, you see how you can create your own user input functionality in a standardized context using `FieldBlock`. This allows you to display labels, error messages and other surrounding elements in a consistent manner with the ready-made fields found in Eufemia Forms.

> Remember that everything that happens by itself using `useDataValue` and the rest of the available helper functionality, you can override individually to make the component work exactly as you want. If, for example, you need to carry out your own custom validation and cannot use the built-in validation with JSON Schema or by sending in a derivative validator as is done in the example above, you can write your own logic for this, and send the result in as props to `FieldBlock` in the form of `error`. All direct props override standard handling, so you have full control over your component. If you need something that looks even more different than the usual fields, you can drop `FieldBlock` and display surrounding elements in other ways if your project requires it, but still get all the help of data flow logic like `useDataValue` and other help functions in Eufemia Forms offer.

Reviewer: ```suggestion
Remember that everything that happens by using `useDataValue` and the rest of the available helper functionality, you can override the behaviour individually to make the component work exactly as you want. 

If – for example; you need to carry out your own custom validation and cannot use the built-in validation with a JSON Schema or by sending in a derivative validator (as is done in the example above), you can write your own logic for it, and send the result in as props to `FieldBlock` in the form of `error`. All direct props override standard handling, so you have full control over your component. 

If you need something that looks even more different than the usual fields, you can drop `FieldBlock` and display surrounding elements in other ways – but still get all the help of a data flow logic, such as `useDataValue` offers.
```
